### ISSUE TITLE
~sircmpwn/sr.ht-docs: tutorials/builds.sr.ht/github-integration.md - sourcehut git

### ISSUE THREAD

**Issue by tg-z** [tg-z/web-clip#979]
* * *

### title: Integrating [builds.sr.ht](http://builds.sr.ht/ "Linkify Plus Plus") with GitHub

This is an adaptation of our [getting started with builds.sr.ht](https://git.sr.ht/~sircmpwn/sr.ht-docs/tree/master/tutorials/getting-started%20with%20builds.sr.ht) tutorial, but for code hosted on GitHub.

### Build manifests

Unlike platforms like Jenkins, [builds.sr.ht](http://builds.sr.ht/ "Linkify Plus Plus") does not allow you to pre-configure jobs. And unlike platforms like Travis, jobs are not inherently tied to a git repository. Each job on [builds.sr.ht](http://builds.sr.ht/ "Linkify Plus Plus") is described ad-hoc with a build manifest, which can be submitted to [builds.sr.ht](http://builds.sr.ht/ "Linkify Plus Plus") for processing.

Let's start with a basic manifest:

image: alpine/edge
tasks:
- example: |
    echo "hello world"

This is a build manifest, written in [YAML](http://yaml.org/). When we submit this to [builds.sr.ht](http://builds.sr.ht/ "Linkify Plus Plus"), it will boot up an [Alpine Linux](https://alpinelinux.org/) virtual machine using the edge release of Alpine Linux. Then it will execute each of our build tasks - in this case, saying hello world.

### Submitting jobs on the web

[builds.sr.ht](http://builds.sr.ht/ "Linkify Plus Plus") has a web submission form, where you can paste a build manifest and submit the job without any additional configuration. This is a useful way of testing build manifests before giving them a permanent home, or running one-off tasks. Visit the [job submission form](https://builds.sr.ht/submit) and paste in the example manifest. Add a note, perhaps my first job, and click submit to run the job.

You'll be redirected to the job detail page. In a moment, one of our job runners will pick up the task and start processing it. Within a few seconds, you should see hello world shown under the example task.

### Adding git repositories to builds

Let's try building [mrsh](https://github.com/emersion/mrsh), which depends on [meson](https://mesonbuild.com/). Here's a build manifest for it:

image: alpine/edge
packages:
- meson
sources:
- https://github.com/emersion/mrsh
tasks:
- configure: |
    cd mrsh
    meson build
- build: |
    cd mrsh
    ninja -C build
- test: |
    cd mrsh
    ninja -C build test

Before starting your tasks, [builds.sr.ht](http://builds.sr.ht/ "Linkify Plus Plus") will clone each repository listed in sources to the build environment. You can have as many or as few (including zero) git repositories as you like. [builds.sr.ht](http://builds.sr.ht/ "Linkify Plus Plus") will also install [Alpine Linux's meson package](https://pkgs.alpinelinux.org/package/edge/main/x86_64/meson) before starting your build. This uses Alpine's native `apk` package manager - other images use different package managers.

### Testing on other platforms

Portability is important - so let's try running the same manifest on another operating system.

image: freebsd
packages:
- meson
sources:
- https://github.com/emersion/mrsh
tasks:
- configure: |
    cd mrsh
    meson build
- build: |
    cd mrsh
    ninja -C build
- test: |
    cd mrsh
    ninja -C build test

This one happens to work without any changes, but note that some images have different names for packages, different distributions of coreutils, and so on.

### Adding these builds to your GitHub repository

Try making a new mrsh repository on your GitHub account. Note that forks won't work - so make sure you make a _new_ repository and push the mrsh code to it. Take a look at the `.builds` directory in mrsh: each of these build manifests can be submitted on push or pull request by rigging up a [dispatch.sr.ht](http://dispatch.sr.ht/ "Linkify Plus Plus") task.

Go to [dispatch.sr.ht](https://dispatch.sr.ht/) and Configure new task. Pick GitHub commits > [builds.sr.ht](http://builds.sr.ht/ "Linkify Plus Plus") jobs and click Add task for your new mrsh repository. That's all you have to do! Now let's make a dummy commit and push it to GitHub to test it out:

Head over to your [builds.sr.ht dashboard](https://builds.sr.ht/) and you should see your build begin momentarily!

### Testing pull requests on GitHub

If you want to run your CI against pull requests on GitHub, follow a similar procedure, but select GitHub pull requests > [builds.sr.ht](http://builds.sr.ht/ "Linkify Plus Plus") jobs instead. Then, each new pull request that comes into your repo will be built on [builds.sr.ht](http://builds.sr.ht/ "Linkify Plus Plus") and the pull request status updated with the build results.

### Why doesn't my GitHub repo show up?

There are a couple of limitations:

*   Forks are not supported
*   You must have admin access to the repo (test this by trying to add a webhook through the GitHub UI manually)

If neither of these are the issue, [write us an email](mailto:sir@cmpwn.com).

* * *

Other resources:

*   [builds.sr.ht user manual](https://git.sr.ht/builds.sr.ht)
*   [Build manifest reference](https://git.sr.ht/builds.sr.ht/manifest.md)
*   [dispatch.sr.ht](https://git.sr.ht/dispatch.sr.ht) 
 https://git.sr.ht/~sircmpwn/sr.ht-docs/tree/master/tutorials/builds.sr.ht/github-integration.md
